### ISSUE TITLE
Add parallel attention layers and Multi-Query Attention (MQA) from PaLM to the fast path for transformers

### ISSUE THREAD

**Issue by ani300** [pytorch/pytorch#95210]
### 🚀 The feature, motivation and pitch

Parallel attention layers and MQA are introduced in PaLM [1]. The “standard” encoder layer, as currently implemented in PyTorch, follows the following structure:

Serial attention (with default `norm_first=False`):

`z = LN(x + Attention(x)); y = LN(z + MLP(z))`

A parallel attention layer from PaLM, on the other hand, is implemented as follows:

Parallel attention (with default `norm_first=False`):
`y = LN(x + Attention(x) + MLP(x))`

Parallel attention (with `norm_first=True`):

`y = x + Attention(LN(x)) + MLP(LN(x))`

As for MQA, the description from [1] is pretty concise and explains the advantage as well: “The standard Transformer formulation uses k attention heads, where the input vector for each timestep is linearly projected into “query”, “key”, and “value” tensors of shape [k, h], where h is the attention head size. Here, the key/value projections are shared for each head, i.e. “key” and “value” are projected to [1, h], but “query” is still projected to shape [k, h]. We have found that this has a neutral effect on model quality and training speed (Shazeer, 2019), but results in a significant cost savings at autoregressive decoding time. This is because standard multi-headed attention has low efficiency on accelerator hardware during auto-regressive decoding, because the key/value tensors are not shared between examples, and only a single token is decoded at a time.”

In our own experiments, we have seen a 23% speedup in training for T5-large when using parallel attention, with the same convergence, due to the ability to fuse the attention and feed-forward projection layers into a single matmul operation. Furthermore, adding MQA on top shaves off an extra 10% training time. See the image below for our training loss over time.

![loss_curve](https://user-images.githubusercontent.com/919977/220369662-0fb39569-bfa2-411d-b03c-bbac8f6737da.png)

Of course, this is not free, and we have found that for larger models (3B and 11B parameters), memory can peak high enough that it can OOM an A100 GPU, even when running with FSDP.

As we build larger models and optimize them for inference, we would like to take advantage of the fast path for Transformers already available for the serial implementation, given the already observed advantage in training and the claimed speed-up on inference from using MQA, as well as explore ways to make this approach more memory efficient in PyTorch.

We envision two ways in which this can happen:

1. Through new parameters in both nn.TransformerEncoderLayer and nn.TransformerDecoderLayer (e.g. booleans `use_parallel_attention` and `use_mqa`), that would route the code through the parallel path and/or MQA.

1. New layers nn.TransformerParallelEncoderLayer and nn.TransformerParallelDecoderLayer

Both paths would eventually call the native `transformer_encoder_layer_forward` (https://github.com/pytorch/pytorch/blob/0dceaf07cd1236859953b6f85a61dc4411d10f87/aten/src/ATen/native/transformers/transformer.cpp#L65) function (or a new parallel version of it that ensures kernel fusing) if all conditions to hit the fast path are met.

cc: @daviswer @supriyogit @raghukiran1224 @mudhakar @mayank31398

**Comment by cpuhrsch** [comment]
Thank you for posting this @ani300! We'll discuss it internally. It seems like a good application of scaled_dot_product_attention and torch.compile. cc @drisspg

**Comment by cpuhrsch** [comment]
In particular for `y = x + Attention(LN(x)) + MLP(LN(x))` you should be able to already do quite a bit of horizontal fusion manually by using `chunk`. For example see how we pack three separate matmuls into one within [sdp.py](https://github.com/pytorch/pytorch/blob/master/benchmarks/transformer/sdp.py#L110-L118).

**Comment by vgoklani** [comment]
SantaCoder was another model leveraging Multi-Query Attention:

https://huggingface.co/bigcode/santacoder

We have a pure PyTorch implementation, but would love an optimized version. We already use F.scaled_dot_product_attention() for training.

**Comment by HamidShojanazeri** [comment]
@vgoklani thanks for sharing, wonder if you have observed any speed up in your training using `F.scaled_dot_product_attention() `

**Comment by vgoklani** [comment]
Hello @HamidShojanazeri - we're seeing a ~5% increase in performance when using F.scaled_dot_product_attention(), where both versions are using PT2 and we are compiling the models. The former does use 14% less ram, however, it crashes when we increase the batch-size :) not sure yet what to make of that.

**Comment by cpuhrsch** [comment]
@vgoklani - Can you share more details on this crash? Is this on a most recent nightly? cc @drisspg

**Comment by vgoklani** [comment]
Hi @cpuhrsch - I just updated to the latest nightly (2.0.0.dev20230223+cu118).  There are no more crashes.

Here are the stats (grad-accumulation disabled for all cases);

1. pytorch compiled + no flash + batch_size=32
2.18 it/s
GPU RAM: 44646MiB / 49140MiB

2. pytorch compiled + flash enabled + batch_size=32
2.29 it/s
GPU RAM: 38496MiB / 49140MiB

3, pytorch compiled + flash enabled + batch_size=40
1.85 it/s
GPU RAM: 47656MiB / 49140MiB

Both case 2 & 3 are processing the same amount of data (i.e. 2.29 * 32 ~ 1.85 * 40), so no crashes, but no real gain either.

I am running this on an NVIDIA A6000 card (Ampere/48GB of GPU RAM)

**Comment by cpuhrsch** [comment]
@vgoklani Depending on the exact details of your model architecture (for example it might be dominated by the feedforward network and not the attention calculation) and how much you're using existing, already optimized modules (like nn.MHA) the gains might not be significant. But I'm happy to see that you're getting some gains at least, including much decreased memory footprint and more it/s.  The slowdown you're seeing for batch_size=40 might be explained by the GPU memory allocator. To really dig into the mile optimizations here we'll need to look into a GPU profile of the workload.

**Comment by vgoklani** [comment]
thanks for the response @cpuhrsch 
What do you mean exactly by " a GPU profile of the workload"? Is this something I could see specifically? I have been thinking about profiling the code, but I'm unsure where/how to start. Is there a good reference for profiling pytorch code? thanks!

**Comment by cpuhrsch** [comment]
@vgoklani - I recommend using https://developer.nvidia.com/nsight-systems via the [`nsys`](https://docs.nvidia.com/nsight-systems/UserGuide/index.html) command line tool. This can capture GPU traces for any arbitrary CUDA workload.

Otherwise a more accessible approach might to use the [PyTorch profiler](https://pytorch.org/tutorials/recipes/recipes/profiler_recipe.html#using-tracing-functionality).

**Comment by HamidShojanazeri** [comment]
@vgoklani thanks a lot for sharing the details on the perf gains and the issue, lets open another ticket on that to keep the focus here on parallel attention.

**Comment by sentialx** [comment]
How did you manage to get scaled_dot_product_attention to work with MQA? I get shape mismatch errors

**Comment by fxmarty** [comment]
Hi, it appears PyTorch SDPA can not dispatch to mem-efficient attention or flash-attention for the shapes:

```
query = [batch_size, num_heads, query_length, head_dim]
key = [batch_size, 1, past_length, head_dim]
value = [batch_size, 1, past_length, head_dim]
```

**Comment by fxmarty** [comment]
Yes that's what I did!

**Comment by fxmarty** [comment]
related https://github.com/pytorch/pytorch/issues/112577
